scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala

```diff
@@ -125,101 +119,164 @@ class FormatWriter(formatOps: FormatOps) {
 
   def getFormatLocations(
       toks: Array[FormatToken],
-      splits: Vector[Split],
+      state: State,
       debug: Boolean
-  ): Array[FormatLocation] = {
-    require(toks.length >= splits.length, "splits !=")
-    val statesBuilder = Array.newBuilder[FormatLocation]
-    statesBuilder.sizeHint(toks.length)
-    var currState = State.start
-    splits.zipWithIndex.foreach {
-      case (split, i) =>
-        val tok = toks(i)
-        currState = State.next(currState, styleMap.at(tok), split, tok)
-        statesBuilder += FormatLocation(tok, split, currState)
-        // TIP. Use the following line to debug origin of splits.
-        if (debug && tokens.length < 1000) {
-          val left = cleanup(tok.left).slice(0, 15)
-          logger.debug(
-            f"$left%-15s $split ${currState.indentation} ${currState.column}"
-          )
-        }
+  ): FormatLocations = {
+    require(toks.length >= state.depth, "splits !=")
+    val result = new Array[FormatLocation](state.depth)
+    var curState = state
+    while (curState.depth != 0) {
+      val idx = curState.depth - 1
+      val tok = toks(idx)
+      result(idx) = FormatLocation(tok, curState)
+      curState = curState.prev
+      if (debug && toks.length < 1000) {
+        val left = cleanup(tok.left).slice(0, 15)
+        logger.debug(
+          f"$left%-15s ${curState.split} ${curState.indentation} ${curState.column}"
+        )
+      }
     }
-    statesBuilder.result()
+    new FormatLocations(result)
   }
 
-  /**
-    * Reconstructs path for all tokens and invokes callback for each token/split
-    * combination.
-    */
-  def reconstructPath(
-      toks: Array[FormatToken],
-      splits: Vector[Split],
-      debug: Boolean
-  )(
-      callback: (State, FormatToken, String, Map[FormatToken, Int]) => Unit
-  ): Unit = {
-    require(toks.length >= splits.length, "splits !=")
-    val locations = getFormatLocations(toks, splits, debug)
-    val tokenAligns = alignmentTokens(locations)
-    var lastModification = locations.head.split.modification
-    locations.zipWithIndex.foreach {
-      case (FormatLocation(tok, split, state), i) =>
-        val previous = locations(Math.max(0, i - 1))
+  class FormatLocations(val locations: Array[FormatLocation]) {
+
+    val tokenAligns: Map[TokenHash, Int] = alignmentTokens(locations)
+
+    def iterate: Iterator[Entry] =
+      Iterator.range(0, locations.length).map(new Entry(_))
+
+    private def getAlign(tok: Token, alignOffset: Int = 0): Int =
+      tokenAligns.get(hash(tok)).fold(0)(_ + alignOffset)
+
+    class Entry(val i: Int) {
+      val curr = locations(i)
+      val previous = locations(math.max(i - 1, 0))
+
+      @inline def tok = curr.formatToken
```

Comment:
I don't sure that inline optimization enabled for scalafmt builds

Reply:
i understand. well, in case it is enabled in the future, should i keep these annotations?

Reply:
i understand. well, in case it is enabled in the future, should i keep these annotations?

Reply:
Yes, of course. Just noticing that now it will not be inlined